src/Microsoft.Security.Utilities.Core/Detections.cs

```diff
@@ -0,0 +1,90 @@
+        /// DevOps personal access token (PAT).'". A portion of the plaintext
+        /// finding is preserved to introduce uniqueness in the output and to
+        /// assist in searching the textual scan target in which the data was
+        /// found.
+        /// </summary>
+        /// <param name="detection">A detection originating in a textual scan
+        /// target.</param>
+        /// <param name="input">The textual input that was scanned.</param>
+        /// <returns>A message with a partial rendering of the plaintext
+        /// finding, suitable for writing to the console, an IDE error list,
+        /// etc.</returns>
+        public static string Format(Detection detection, string input)
+        {
+            string match = input.Substring(detection.Start, detection.Length);
+            string truncated = TruncateSecret(match);
+
+            bool isHighConfidence = detection.Metadata.HasFlag(DetectionMetadata.HighConfidence);
+
+            string verb = isHighConfidence ? "is" : "may comprise";
+
+            string suffix = !string.IsNullOrEmpty(detection.CrossCompanyCorrelatingId)
+                ? $" The correlating id for this detection is {detection.CrossCompanyCorrelatingId}."
+                : string.Empty;
+            
+            return $"'{truncated}' {verb} {detection.Label}.{suffix}";
+        }
+
+        /// <summary>
+        /// Truncates a string to a specified length, adding an ellipsis if any
+        /// of the input string is removed. This helper will remove up to two
+        /// trailing equals signs from the input string before truncating it.
+        /// This is intended to help cover the standard case where a secret is
+        /// base64-encoded data.
+        /// </summary>
+        /// <param name="text">The text to truncate.</param>
+        /// <param name="lengthExclusiveOfEllipsis">The desired length of the
+        /// truncated text, exclusive of the ellipsis, if added.</param>
+        /// <returns>The rightmost truncated contents of the strength or the
+        /// entire string (if its length is equal to or less than the specified
+        /// length).</returns>
+        public static string TruncateSecret(string text, int lengthExclusiveOfEllipsis = 6)
+        {
+            text ??= string.Empty;
+            string truncatedText = text.Replace("-", "+").Replace("_", "/");
+
+            try
+            {
+                // If we receive valid base64 or url-safe base64, we will
+                // elide any trailing equal signs, so that the padding that's
+                // retained in the truncated strength does not apply to the 
+                // length constraint. 
+                Convert.FromBase64String(truncatedText);
+                truncatedText = text.TrimEnd('=');
+            }
+            catch (FormatException)
+            {
+            }
+
+            string suffix = new string('=', text.Length - truncatedText.Length);
```

Comment:
This first-chance exception might get annoying when debugging. I think we could do what the new description literally says and remove up to two equals regardless of valid base64 or not.

Something like this?

```C#
string suffix = string.Empty;
if (text.EndsWith("==", System.StringComparison.Ordinal))
{
    suffix = "==";
}
else if (text.EndsWith('='))
{
    suffix = "=";
}

string truncatedText = text.Substring(0, text.Length - suffix.Length);
```

Reply:
I'd like to ship what we have. Here's why, I actually like the fact that we only attempt this operation for valid base64 strings. If I took your suggestion a bunch of thoughtful tests would begin to fail. :)

What I suggest is that I go back in later and add a more developed static 'IsValidBase64' check (rather than inferring this from the fact that Convert.FromBase64String doesn't raise an exception. #187

Reply:
Sounds good.